Unit configuration

matteocelani opened the pull request:
## Description
This pull request updates the unit of measurement for the contract size in the project from Kibibyte to Kilobyte. This change was made to align the unit of measurement with the official Ethereum documentation, which expresses contract size in Kilobytes:
-  [Downsizing Contracts To Fight The Contract Size Limit](https://ethereum.org/en/developers/tutorials/downsizing-contracts-to-fight-the-contract-size-limit/)
- [Hard Fork No. 4: Spurious Dragon](https://blog.ethereum.org/2016/11/18/hard-fork-no-4-spurious-dragon)

The change was made to improve console readability and ensure data compatibility with the unit of measurement used in the official documentation.

ItsNickBarry commented:
I wouldn't consider those "official documentation."  One is a tutorial and the other is a blog.  The blog doesn't actually mention KB - it describes the limit as "24576 bytes."  The [EIP](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-170.md) defines `MAX_CODE_SIZE` as `0x6000`.

The lack of consistency in general has created a lot of confusion.  I've often seen developers incorrectly describe the limit as "24 KB" as a result.  Because "24" is much more memorable and concise than "24.576," I've decided to denominate the limit in KiB.

I would accept this PR if it added a configuration option to display sizes in KB, but I'm going to keep KiB as the default.

matteocelani commented:
I referred to the official documentation as, by searching on the official [Ethereum](https://ethereum.org/en/) website, one can find out that the maximum size of a contract is 24576 bytes or 24,576 KB (expressed in decimal units).

Furthermore, since 24.576 KB is equal to 23.4375 KiB, it can initially be confusing to have a contract that appears to be below the limit of "24" but has already exceeded it.

## Changes Made

In any case, to align with the requested changes and avoid confusion, I added the ability to configure the unit of measurement, specifically:

- Added a configuration option to select the unit of measurement for contract size.
- Set the default value to kibibytes.
- Created the option to express the data in bytes or kilobytes.

ItsNickBarry commented on index.d.ts:
```diff
@@ -10,6 +10,7 @@ declare module 'hardhat/types/config' {
       only?: string[],
       except?: string[],
       outputFile?: string,
+      unit?: 'b' | 'kib' | 'kb',
```
Do you know of a way to make the type case-insensitive?

ItsNickBarry replied:
If not, let's use appropriate capitalization here: "B", "KiB", and "kB"

ItsNickBarry commented on tasks/size_contracts.js:
```diff
@@ -25,6 +19,16 @@ task(
   const config = hre.config.contractSizer;
 
+  const SIZE_LIMIT = 24576;
+  const UNITS = config.units == 'b' ? 'B' : config.units == 'kb' ? 'Kb' : 'KiB';
```
"B" should be capitalized.  "k" should not.  It seems that it's conventional in computer science to capitalize the "k", but this is disallowed by SI.

```suggestion
  const UNITS = config.units == 'b' ? 'B' : config.units == 'kb' ? 'kB' : 'KiB';
```

ItsNickBarry replied:
The `UNITS` variable can probably be removed in favor of just using `config.units`.  See comment in `index.d.ts`.

matteocelani replied:
Having made the type case-insensitive, I would leave the `UNIT` variable so that it can be formatted correctly in console.

ItsNickBarry reviewed (changes requested):
> Furthermore, since 24.576 KB is equal to 23.4375 KiB, it can initially be confusing to have a contract that appears to be below the limit of "24" but has already exceeded it.

Not exactly.  23.4375 KiB is equal to 24 kB, which does not exceed the limit.  The limit is 24575 B = 25.576 kB = 24 KiB.

matteocelani commented:
Ok, great! When searching online, it's easy to get confused.

24576 B correspond to 24.576 kilobytes (kB) and 24 kibibytes (KiB).

```
KiB = B / 1024
kB = B / 1000
```

```
KiB = 24576 B / 1024 = 24 KiB
kB = 24576 B / 1000 = 24.576 kB
```

## Changes Made

- made the `unit` type case-insensitive
- formatted the `UNIT` of measurement correctly before printing to the console.

ItsNickBarry reviewed (approved):
Looks good, I'm going to merge.  However, I'm having second thoughts about case-insensitive input.  Theoretically, someone could enter "kb" and expect output in kilobits, but would receive output in kilobytes ("kB").  If I decide to make changes to this, I will do so on a new branch and merge before publishing.

ItsNickBarry commented:
Thanks for the contribution.

ItsNickBarry commented:
Published in `v2.9.0`.
